### PR TITLE
support partial aggregation and masking for AverageEpisodicAggregationMetric

### PR DESCRIPTION
**Pull request by hnyu**
Sometimes the fields of env info don't always appear in every episode. For example, we randomly sample a different task for each episode and the task env info fields might be different for different episodes.

However, our metric class require a complete env info spec (that contains all the possible fields) during the initialization, and it assumes that each field will be accumulated for every episode. This means that usually we need to fill in dummy metric values for those fields that are actually not active to be compatible. 

This PR makes `AverageEpisodicAggregationMetric` more flexible by allowing the user to skip aggregating certain metric fields by providing "invalid" values. Basically, an invalid value tells the accumulator not to aggregate the current value. For any field, if the values of an entire episode are invalid, then at the episode end, the buffer won't be pushed into a new accumulated value, and the buffer's mean stays the same.

### CODE REVIEW

File: alf/metrics/metrics.py [hunk 1]
```diff
@@ -230,40 +238,54 @@ def call(self, time_step):
 
         is_first = time_step.is_first()
 
-        def _update_accumulator_(path, acc, val):
-            """In-place update of the accumulators."""
+        def _update_accumulator_(path, mask, acc, val):
+            """In-place update of the accumulators and mask."""
+            val_valid = torch.isfinite(val)
+            # If at any step the value is valid, then the acc value becomes valid
```
Comment (emailweixu) [inline alf/metrics/metrics.py:253]: This behavior should be clearly document in the doc-string of `AverageEpisodicAggregationMetric`

File: alf/metrics/metrics.py [hunk 2]
```diff
@@ -230,40 +238,54 @@ def call(self, time_step):
...
+            mask[:] = torch.where(is_first, torch.zeros_like(mask),
+                                  mask + val_valid)
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: `mask | val_valid`

File: alf/metrics/metrics.py [hunk 3]
```diff
@@ -230,40 +238,54 @@ def call(self, time_step):
...
 
-        def _episode_end_aggregate_(path, buf, acc):
+        def _episode_end_aggregate_(path, mask, buf, acc):
             value = self._extract_and_process_acc_value(
                 acc, last_episode_indices)
             # If the metric's name ends with '@step', the value will
             # be further averaged over episode length so that the
             # result is per-step value.
             if path.endswith('@step'):
                 value = value / self._current_step[last_episode_indices]
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: Should `self._current_step[last_episode_indices]` take into val_valid into account?

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/metrics/metrics.py (hunk 1 above)
Comment (hnyu) [reply alf/metrics/metrics.py:253]: > This behavior should be clearly document in the doc-string of `AverageEpisodicAggregationMetric`

Added.

File: alf/metrics/metrics.py (hunk 3 above)
Comment (hnyu) [reply alf/metrics/metrics.py]: > Should `self._current_step[last_episode_indices]` take into val_valid into account?

Updated to have field-dependent steps. Still keep the original `self._current_step` as it is needed by a subclass.

**Review by emailweixu — APPROVED** [review]
